apps/hash-graph/lib/graph/src/store/fetcher.rs

```diff
@@ -131,3 +109,213 @@ where
         &mut self.store
     }
 }
+
+#[async_trait]
+impl<S, A, R: Record> Read<R> for FetchingStore<S, A>
+where
+    A: Send + Sync,
+    S: Read<R> + Send,
+{
+    async fn read(
+        &self,
+        query: &Filter<R>,
+        time_projection: &TimeProjection,
+    ) -> Result<Vec<R>, QueryError> {
+        self.store.read(query, time_projection).await
+    }
+}
+
+#[async_trait]
+impl<S, A> AccountStore for FetchingStore<S, A>
+where
+    S: AccountStore + Send,
+    A: Send + Sync,
+{
+    async fn insert_account_id(&mut self, account_id: AccountId) -> Result<(), InsertionError> {
+        self.store.insert_account_id(account_id).await
+    }
+}
+
+#[async_trait]
+impl<S, A> DataTypeStore for FetchingStore<S, A>
+where
+    S: DataTypeStore + Send,
+    A: Send + Sync,
+{
+    async fn create_data_types(
+        &mut self,
+        data_types: impl IntoIterator<
+            Item = (DataType, impl Borrow<OntologyElementMetadata> + Send + Sync),
+            IntoIter: Send,
+        > + Send,
+    ) -> Result<(), InsertionError> {
+        self.store.create_data_types(data_types).await
+    }
+
+    async fn get_data_type(
+        &self,
+        query: &StructuralQuery<DataTypeWithMetadata>,
+    ) -> Result<Subgraph, QueryError> {
+        self.store.get_data_type(query).await
+    }
+
+    async fn update_data_type(
+        &mut self,
+        data_type: DataType,
+        actor_id: UpdatedById,
+    ) -> Result<OntologyElementMetadata, UpdateError> {
+        self.store.update_data_type(data_type, actor_id).await
+    }
+}
+
+#[async_trait]
+impl<S, A> PropertyTypeStore for FetchingStore<S, A>
+where
+    S: PropertyTypeStore + Send,
+    A: Send + Sync,
+{
+    async fn create_property_types(
+        &mut self,
+        property_types: impl IntoIterator<
+            Item = (
+                PropertyType,
+                impl Borrow<OntologyElementMetadata> + Send + Sync,
+            ),
+            IntoIter: Send,
+        > + Send,
+    ) -> Result<(), InsertionError> {
+        self.store.create_property_types(property_types).await
```

Comment:
By the way, what does this delegate to at the end? Is it passed on to the postgres store, or what happens? And if it is passed to postgres, how does that happen in the router (my questions are likely answered in your next PR in the stack :) )

I read some stuff about using semigroups to model the routing chain of a HTTP request and I suppose dependencies could work in a sort of similar way for us, but honestly I have no clue where we would even begin to leverage something like that. Just seems like a powerful, existing pattern that would be highly composable.

Reply:
Yes, it will still be passed down to the Postgres store, only some logic will be added in the follow-up to also insert other ontology types before.